### ISSUE TITLE
Port to Docker

### ISSUE THREAD

**Issue by frob** [geerlingguy/drupal-vm#139]
In #5 you said you might port this over to Docker.

I would like to help with this. I just had a few questions. I am new to Docker but this looks like the way to go with this. 

How would you plan on making this work with Docker? I was thinking make it inherit the ubuntu container then install ansible and then run the ansible provision. Is that what you had in mind?

**Comment by geerlingguy** [comment]
That's one way to do it, but it wouldn't really give us much advantage over the current setup, since everything would be in one monolithic Docker container. I would rather take a longer-term, more holistic approach, trying to follow Docker best practices, and making the entire project much more maintainable—especially in a way that, like Drupal VM currently does, allows for the same/similar configuration for local development and cloud prod/staging environments.

I'm going to keep this particular issue closed for now, but feel free to continue work on experimenting with Docker—even if it's only something you build and toss away, it's a good experience to work with Docker and figure out where things get tricky, what things it makes easier, etc. If you need some extra help, there are a few bits on Docker in [Ansible for DevOps](http://ansiblefordevops.com/) to get you started :)

And as a final note, there's a chance that other projects like Kalabox and DevShop (which are already using containers) may someday make for a better solution than Drupal VM, but we'll see...

**Comment by geerlingguy** [comment]
(marked wontfix for now... but it is a long term goal)

**Comment by frob** [comment]
I think I will start with the monolithic approach. From what I understand it will still make better use of resources being in a container, especially if multiple containers are being used.

**Comment by bwowk** [comment]
Any updates on this?

**Comment by geerlingguy** [comment]
@bwowk - Every time I start into a monster blog post explaining why Drupal VM isn't running on/in Docker yet, I end up with at least 1,000 words but not good conclusion.

Since the foundational bits have continued to shift... and even 1.12 hasn't been as stable/good-across-all-platforms as I'd hoped, I'm still taking a 'wait-and-see' approach.

At this time, I've played extensively with Kalabox, Drocker, Docker4Drupal, the official Drupal image, and about a dozen other community-maintained Docker solutions for Drupal, and there are parts of the UX/DX that are extremely grating still.

As a make-a-config-for-a-particular-site solution, Docker is at a point where—for a team with at least one person who can devote time to 'Devopsy' stuff—it can be more efficient/productive than something like Drupal VM. But for generalized use, especially for teams that are cross-platform and need easy 'push-button' setup, Docker is far behind where I think it needs to be.

Things like customization of packages inside containers (e.g. this project needs PHP's `bz2` extension, while this other one needs `ssh2`), intra-container networking (still not rock solid like I want it to be), persistence with databases and other files, automated management of hosts files (for friendly domains rather than random ports like 24892)/host networking, and even performance variation for things like mounted volumes (especially on Mac) mean I would spend at least 2-4x more time _helping with support problems_ rather than fixing actual bugs or improving Drupal VM.

I've spent a _lot_ more time with Docker in the past year. And there are some amazing things about Docker, Kubernetes, and other containery things. But I don't feel like they've hit they're 'Vagrant' moment, where things came together to make the _common_ developer's life easier/simpler. It still requires too much handholding and little things here and there which require more devoted time to work on the _tool_ rather than the things being _built with the tool_ (e.g. Drupal sites).

**Comment by bwowk** [comment]
Thanks for the prompt response :) I just asked because I was curious if there was some ongoing effort in this sense.

Yeah I definitely agree that Docker is only stable-ish on a cross-platform standpoint, and there's a lot of ground to cover on providing a uniform experience across platforms. To make a cross platform setup we'd have to make do without the best bits of the docker tools, because there's no denying it's a Linux-first tool.

**Comment by Vacilando** [comment]
@bwowk: hmm, I'd rephrase it to "it was on Linux first". The new Docker for Windows (albeit only for Windows 10 Pro and up) simply rocks. For the first time can Windows users enjoy the simplicity and versatility of Docker and not lose any performance due to working under Windows.

http://docker4drupal.org/ is making an excellent use and business from it. There is almost no competition. I hope Drupal VM / @geerlingguy overcomes the hesitations and strives to build an even better Drupal VM on Docker. Competition like that can only do good!

**Comment by geerlingguy** [comment]
@Vacilando - The main thing is Docker for Mac is atrocious right now, and that's the main reason I can't jump in headfirst. There are comparatively few developers using Linux/Windows, and almost every other tool on the market is Linux/Mac-first. It's such a weird change of events to have Linux/Windows be pretty awesome and Mac support to be _so bad_.

The two major issues for me are horrible `osxfs` performance (think VirtualBox native synced folders, but 5x slower), and the fact that the 'qcow' file is (a) limited to 64 GB in size (which can cause weird issues for Docker power-users), and (b) it doesn't get garbage collected correctly, meaning anyone without a giant SSD will run out of space after downloading dozens or thousands of images (unless they manually clear out the qcow file).

There are issues for both of these, but so far both have been open for months with very little response from the Docker dev team, just a lot of +1s everywhere.

Docker Machine / using a VM to run Docker inside it is a solution, but IMO that's worse than just using a VM directly like with Drupal VM... you have a VM running a bunch of containers, and that's an abstraction too far—at least for the 'plug and play' use case that I'm targeting. I don't want developers to have to learn minutiae about the tools Drupal VM uses—e.g. you might not know _anything_ about Vagrant or VirtualBox (much less Ansible), but you can still build a highly-custom environment in < 10 minutes on a fast internet connection, and access it at a nice URL like `local.mysite.com`, configuring just a few lines of `config.yml` and nothing else.

With Docker, the above is just not possible right now (unless you are deep into the tools and ecosystem).

**Comment by thom8** [comment]
If anyone is interested, I've done a bit of work using Drupal VM roles with docker. 

`docker-compose` - https://github.com/thom8/beetbox-docker
`vagrant` - https://github.com/thom8/beetbox-docker/tree/vagrant

However, early results have only highlighted these terrible performance issues.

*only tested with Docker for Mac.

**Comment by geerlingguy** [comment]
> However, early results have only highlighted these terrible performance issues.

Heh, yep. If Docker resolved https://github.com/docker/for-mac/issues/77, I'd be like 99x more likely to get serious about doing anything with Docker. As it stands, I'm not going to add on _more_ layers of shims and complexity (on top of Docker, xhyve, base images, etc.) to try to get decent performance for Drupal VM users with Macs (the majority, based on my informal polling).

**Comment by fago** [comment]
>Things like customization of packages inside containers (e.g. this project needs PHP's bz2 extension, while this other one needs ssh2),...

We've been using monolithic docker images that are built with Ansible for that. We are doing so now for almost two years and it works fine. If you are interested, you can find them here: https://github.com/drunomics/docker-images

I like keeping things monolithic as it simplifies things for developers compared to the regular docker approach. That, combined with vagrant for easy customization/provision and hosts file updating support, that works very well for us. However, we did not run into the mentioned performance issues as we are using primarily linux environments. (desktop also)

**Comment by geerlingguy** [comment]
Linking for historical reference:

  - [Drupal VM, The Next Generation](https://github.com/geerlingguy/drupal-vm/issues/996)

And @fago, that's the approach that I'm settling on mostly for the issue linked above. Basically, we can find ways to make it so Drupal VM's Ansible configuration can be used to build Docker images more easily (e.g. pick a PHP version, pick extensions, etc. through config.yml, click a button, then you have your customized stack, ready to put on Docker Hub, or a local registry, and shared with the dev team.

The primary difference in this workflow is that you need to have someone build that Docker image and have it available in a repo somewhere, instead of having each developer build things locally. This saves the 10-20 minutes per developer first-time setup that Drupal VM currently takes, but it does require a little more team discipline in getting Docker adoption going.

And if users are on Mac, especially (and somewhat also Windows), there are still a lot of small issues with the Docker app that make things a little more buggy and/or slow than I'd like (even if using machine/toolbox/vagrant).
